base/tuple.jl

```diff
@@ -81,21 +81,60 @@ function _maxlength(t::Tuple, t2::Tuple, t3::Tuple...)
     max(length(t), _maxlength(t2, t3...))
 end
 
-# this allows partial evaluation of bounded sequences of next() calls on tuples,
-# while reducing to plain next() for arbitrary iterables.
-indexed_iterate(t::Tuple, i::Int, state=1) = (@_inline_meta; (getfield(t, i), i+1))
-indexed_iterate(a::Array, i::Int, state=1) = (@_inline_meta; (a[i], i+1))
-function indexed_iterate(I, i)
-    x = iterate(I)
-    x === nothing && throw(BoundsError(I, i))
-    x
-end
-function indexed_iterate(I, i, state)
-    x = iterate(I, state)
-    x === nothing && throw(BoundsError(I, i))
-    x
+# this allows partial evaluation of bounded sequences of iterate() calls on tuples,
+# while reducing to plain iterate() for arbitrary iterables.
+
+arg1(a) = a
+arg1(a, b) = a
+iterate_and_index(t::Tuple) = (arg1, getfield)
+iterate_and_index(t::Array) = (arg1, getindex)
+
+struct BadDestructure
+    a
 end
 
+function destruct_iterate(a)
+    @_inline_meta
+    s = iterate(a)
+    s === nothing && return BadDestructure(a)
+    s
+end
+
+function destruct_iterate(a, b)
+    @_inline_meta
+    s = iterate(a, getfield(b, 2))
+    s === nothing && return BadDestructure(a)
+    s
+end
+
+select_first(a::BadDestructure, i) = throw(BoundsError(a.a, i))
+select_first(a, i) = getfield(a, 1)
+
+iterate_and_index(x) = (destruct_iterate, select_first)
+
+# Nothing is often union'ed into other things. Kill that as quickly as possible
+# to make inference's life easier.
+iterate_and_index(::Nothing) = throw(MethodError(iterate, (nothing,)))
+
+function _rest(itr, state, _)
+    @_inline_meta
+    state === () && return rest(itr)
+    return rest(itr, state[1][2])
+end
+_rest(x::Union{Array,Tuple,NamedTuple,Pair}, _, i) = (@_inline_meta; rest(x, i))
+
+rest(t::Tuple) = t
+rest(t::NTuple{N}, i::Int) where {N} = ntuple(x -> getfield(t, x+i-1), N-i+1)
+rest(a::Array, i::Int=1) = a[i:end]
+# The semantics of `collect` are weird. Better to write our own
+function rest(a::AbstractArray{T}, state...) where {T}
+    v = Vector{T}(undef, 0)
+    # assume only very few items are taken from the front
+    sizehint!(v, length(a))
+    return foldl(push!, Iterators.rest(a, state...), init=v)
+end
+rest(itr, state...) = Iterators.rest(itr, state...)
```

Comment:
@JeffBezanson I would love to have some feedback on this design. It's a bit uglier than I initially imagined, but I don't think we can really avoid that with the way `iterate_and_index` currently works. `base/tuple` is probably also the wrong place for most of this, but I am still not sure, what the best home for `Base.rest` is.

Reply:
The methods can go with their types, like iterate and getindex tend to.

Reply:
Yes, that's what I planned on doing. I just didn't really know where to put the generic `Base.rest` function. `base/tuple.jl` seems like a bit of a weird place for it.